clarified from part 1 to part 4 - After Benjamin's version (#125)
trilitech/tezos-developer-docs: src/pages/tutorials/build-an-nft-marketplace/index.md

### Smart Contracts Overview

There are two contracts for the marketplace.

#### The token contract

On Tezos, FA2 is the standard for non-fungible token contracts. The [template provided by Ligo](https://packages.ligolang.org/package/@ligo/fa) will be used to build out the token contract. The template contains the basic entrypoints for building a Fungible or non-fungible token including:

- Transfer
- Balance_of
- Update_operators

#### Marketplace unique contract

Next, you need to import the token contract into the unique marketplace contract. The latter is bringing missing features as:

- Mint
- Buy
- Sell

## Wine marketplace

The `@ligo/fa` package is a pre-built module from the [Ligo repository](https://packages.ligolang.org/) that provides foundational tools and functions for building and interacting with smart contracts on the Tezos blockchain. 

The next step is to build a wine marketplace extending the `@ligo/fa` package from the [Ligo repository](https://packages.ligolang.org/). 

The goal is to showcase how to extend an existing smart contract and build a frontend on top of it.

> The wine marketplace is adding these features on top of a generic NFT contract :

Reviewer: mentionned earlier line 66 : The next step is to build a wine marketplace extending the `@ligo/fa` package

### What is FA?

Business objects managed by a blockchain are called **assets**. On Tezos you find the term **Financial Asset (abbr. FA)**.

This diagram shows different kinds of assets. Some tangible assets in the real-world, like physical currency and precious metals, are _fungible_, or interchangeable. Other tangible assets, such as specific pieces of art or property, are non-fungible. Similarly, Tezos financial assets can be fungible or non-fungible.

![](http://jingculturecommerce.com/wp-content/uploads/2021/03/nft-assets-1024x614.jpg)

The diagram visually presents the classification of different types of assets based on two main criteria: tangibility and fungibility.

### What is IPFS?

> The InterPlanetary File System is a protocol and peer-to-peer network for storing and sharing data in a distributed file system. IPFS uses content-addressing to uniquely identify each file in a global namespace connecting all computing devices. While Tezos is a blockchain optimized for transactions and smart contracts, it isn't ideal for storing large data, which is where IPFS comes into play. It complements blockchains like Tezos by efficiently handling large file storage. In this tutorial, using the [Pinata](https://www.pinata.cloud/) (free developer plan) , you'll learn to store NFT metadata on IPFS and reference it on Tezos, offering a scalable and cost-effective solution for handling extensive data associated with NFTs. An alternative would be to install a local IPFS node or an API gateway backend with a usage quota.

Reviewer 2: Clarified IPFS: 
> The InterPlanetary File System is a protocol and peer-to-peer network for storing and sharing data in a distributed file system. IPFS uses content-addressing to uniquely identify each file in a global namespace connecting all computing devices. It complements blockchains like Tezos by efficiently handling large file storage. In this tutorial, you use [Pinata](https://www.pinata.cloud/)'s free developer plan to store your NFT metadata on IPFS and reference it on Tezos, demonstrating a scalable and cost-effective solution for handling extensive data associated with NFTs.

{% /callout %}

## Prerequisites

Before building an NFT marketplace, you must install the following tools.

- [npm](https://nodejs.org/en/download/): for managing and installing dependencies in a TypeScript React application.
- [taqueria >= v0.40.0](https://github.com/ecadlabs/taqueria): for Tezos app development and deployment.
- [Docker](https://docs.docker.com/engine/install/): creates containers for consistent app development environments, required by **taqueria**
- [jq](https://stedolan.github.io/jq/download/): command-line tool to parse and manipulate JSON data from **taqueria**
- [Temple wallet](https://templewallet.com/): A browser-based wallet for managing Tezos tokens and interacting with Tezos dApps.
- [`VS Code`](https://code.visualstudio.com/download): A code editor for developing and managing your application's codebase.
- [ligo VS Code extension](https://marketplace.visualstudio.com/items?itemName=ligolang-publish.ligo-vscode): Enhances VS Code with support for the LIGO smart contract language.

The following tools are optional:

> - [`yarn`](https://classic.yarnpkg.com/lang/en/docs/install/#windows-stable): to build and run the front-end (see this article for more details about [differences between `npm` and `yarn`](https://www.geeksforgeeks.org/difference-between-npm-and-yarn/))

Reviewer 2: Moved yarn to required prereqs.

    - To simplify the contract it does not store a counter for the token ID and assumes that the front-end application will keep track of the next ID. Ideally, the contract should manage this ID to avoid conflicts. It is up to you what ID to start at, but usually NFTs start with the ID 0. Optionally, you can also add code to reuse the IDs from burned NFTs.
    - Most of the fields are optional except `decimals` that is set to `0`. A unique NFT does not have decimals, it is a unit
    - By default, the `quantity` for an NFT is `1`, that is why every bottle is unique and there is no need to set a total supply on each NFT.
    - There is no way to tell how many NFTs are in the collection from the contract because the big_map data type does not have a function to return its length. You could add an additional element in the storage to remember the number of NFTs. Another way to get the size of the collection is to index it on the front end.

    **Smart contract implementation for this first training is finished, let's prepare the deployment to ghostnet.**

> 1. Compile the `nft.jsligo` file to create a default taqueria initial storage and parameter file

Reviewer 2: Added a step to ensure that Docker Desktop is running.

          enqueueSnackbar(
            "Wine collection is minting ... it will be ready on next block, wait for the confirmation message before minting another collection",
            { variant: "info" }
          );

          await op.confirmation(2);

          enqueueSnackbar("Wine collection minted", { variant: "success" });

          refreshUserContextOnPageReload(); //force all app to refresh the context
        }
      } catch (error) {
        console.table(`Error: ${JSON.stringify(error, null, 2)}`);
        let tibe: TransactionInvalidBeaconError = new TransactionInvalidBeaconError(
          error
        );
        enqueueSnackbar(tibe.data_message, {
          variant: "error",
          autoHideDuration: 10000,
        });
      }
    };
    ```

    > Note : organize/fix duplicated import declarations if necessary

    ![mint form](/images/mintForm.png)

    Explanations:

    - on Mint button click, upload a file and then call the **pinata API** to push the file to **IPFS**. It returns the hash
    - hash is used in two different ways
      - https pinata gateway link (or any other ipfs http viewer)
      - ipfs link for the backend thumbnail url

>     - TZIP standard requires storing data in `bytes`. As there is no Michelson function to convert string to bytes (using Micheline data PACK is not working, as it alters the final bytes), do the conversion using `char2Bytes` on the frontend side

Reviewer 2: - No need to bold pinata or ipfs
- These bullets really need to be clarified -- there's a lot of info here that is being introduced for the first time

    > Note : Finally, if you remember on the backend, token_id increment management was done in the ui, so you can write this code. It is not a good security practice as it supposes that the counter is managed on frontend side, but it is ok for demo purpose.

> 3. Add this code inside your `MintPage` Component function , every time you have a new token minted, you increment the counter for the next one

Reviewer 2: ```suggestion
3. Add this code inside the `MintPage` component function. Each time it mints a token, it increments the counter for the next token's ID.
```

                    </Card>
                  )
                )}
              </SwipeableViews>
              <MobileStepper
                variant="text"
                steps={Array.from(nftContratTokenMetadataMap!.entries()).length}
                position="static"
                activeStep={activeStep}
                nextButton={
                  <Button
                    size="small"
                    onClick={handleNext}
                    disabled={
                      activeStep ===
                      Array.from(nftContratTokenMetadataMap!.entries()).length - 1
                    }
                  >
                    Next
                    <KeyboardArrowRight />
                  </Button>
                }
                backButton={
                  <Button
                    size="small"
                    onClick={handleBack}
                    disabled={activeStep === 0}
                  >
                    <KeyboardArrowLeft />
                    Back
                  </Button>
                }
              />
            </Box>
    ```

Finally, your imports at beginning of the file should be like this :

>     ```typescript

Reviewer 2: This appears broken in the preview:
<img width="724" alt="Screenshot 2023-10-17 at 3 05 36 PM" src="https://github.com/trilitech/tezos-developer-docs/assets/6494785/36ae8994-79e3-4692-a466-114c986810a6">

    import {
      AddCircleOutlined,
      Close,
      KeyboardArrowLeft,
      KeyboardArrowRight,
    } from "@mui/icons-material";
    import Paper from "@mui/material/Paper";
    import Typography from "@mui/material/Typography";
    import { useFormik } from "formik";
    import React, { useEffect, useState } from "react";
    import * as yup from "yup";
    import { TZIP21TokenMetadata, UserContext, UserContextType } from "./App";
    import { useSnackbar } from "notistack";
    import { BigNumber } from "bignumber.js";
    import { address, bytes, nat } from "./type-aliases";
    import { char2Bytes } from "@taquito/utils";
    import { TransactionInvalidBeaconError } from "./TransactionInvalidBeaconError";
    ```

    and some variables inside your `MintPage` Component function

    ```typescript
    const [activeStep, setActiveStep] = React.useState(0);

    const handleNext = () => {
      setActiveStep((prevActiveStep) => prevActiveStep + 1);
    };

    const handleBack = () => {
      setActiveStep((prevActiveStep) => prevActiveStep - 1);
    };

    const handleStepChange = (step: number) => {
      setActiveStep(step);
    };
    ```

> ## Let's play

Reviewer 2: "Your picture is be pushed to IPFS and displayed." => "Your picture is pushed to IPFS and displayed on the page."

    Install the plugin, then generate a representation of your smart contract objects that writes these files to your frontend app source code.

2. Run the server

    ```bash
    taq install @taqueria/plugin-contract-types
    taq generate types ./app/src
    cd app
    yarn install
    yarn dev
    ```

    > Note: On a **Mac**:green_apple:, sometimes `sed` commands do not work exactly the same as Unix commands. Look at the start script on package.json for Mac below:
    > `   "dev": "if test -f .env; then sed -i '' \"s/\\(VITE_CONTRACT_ADDRESS *= *\\).*/\\1$(jq -r 'last(.tasks[]).output[0].address' ../.taq/testing-state.json)/\" .env ; else jq -r '\"VITE_CONTRACT_ADDRESS=\" + last(.tasks[]).output[0].address' ../.taq/testing-state.json > .env ; fi && vite",`

    The website is ready! You have:

    - last deployed contract address always is refreshed from **taqueria** configuration at each start
    - login/logout
    - the general layout / navigation

    If you try to connect you are redirected to `/` path that is also the wine catalog.

    There are no bottle collections yet, so you have to create the mint page.

> ### Step 2: Mint Page

Author: When I copy the following code, the editor complains about it. Is there any possible to provide the complete Mintpage.tsx code?

              sx={
                isTablet
                  ? { marginTop: "0", marginRight: "0" }
                  : { marginTop: "35px", marginRight: "125px" }
              }
            />
            <Box
              sx={{
                width: isTablet ? "40vw" : "60vw",
                borderColor: "text.secondary",
                borderStyle: "solid",
                borderWidth: "1px",

                height: "calc(100vh - 64px)",
              }}
            >
              <Button
                sx={{
                  position: "absolute",
                  right: "0",
                  display: !formOpen ? "none" : "block",
                }}
                onClick={toggleDrawer(!formOpen)}
              >
                <Close />
              </Button>
            </Stack>
          </form>
        </Box>
      </SwipeableDrawer>

      <Typography variant="h5">Mint your wine collection</Typography>

      {nftContratTokenMetadataMap.size != 0 ? (
        "//TODO"
      ) : (
        <Typography sx={{ py: "2em" }} variant="h4">
          Sorry, there is not NFT yet, you need to mint bottles first
        </Typography>
      )}
    </Paper>
```

> Inside your `MintPage` Component function, all all following elements :

Author: In `MintPage.tsx`? where should I add those elements?
